### PR TITLE
fix: add courseapi

### PR DESCRIPTION
**Pull request by DudeUnleashed**
# Description

Created entities Course. This will allow access to various information relating to courses for the Courseflow component of OnTrack. Various GET, POST, PUT and DELETE endpoints created and implemented, with appropriate tests.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] With the API running, cd into doubtfire-api
- [x] Run the following command: bundle exec rails test test/api/courseflow/course_api_test.rb 
- [x] Can manually test endpoints using postman.

image

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation if appropriate
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have created or extended unit tests to address my new additions
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by sammyrolfe** [comment]
All tests run successfully, all endpoints tested successfully.
Screenshot 2024-08-06 at 8 36 10 pm

File: db/schema.rb [hunk 1]
```diff
@@ -44,7 +44,7 @@
     t.index ["name"], name: "index_campuses_on_name", unique: true
   end
 
-  create_table "comments_read_receipts", charset: "utf8", collation: "utf8_unicode_ci", force: :cascade do |t|
+  create_table "comments_read_receipts", charset: "utf8mb3", collation: "utf8mb3_unicode_ci", force: :cascade do |t|
```
Comment (aNebula) [inline db/schema.rb:47]: How come there's a mixture of `utf8mb3_unicode_ci` and `utf8mb4_general_ci` ?

**Comment by aNebula** [comment]
> All tests run successfully, all endpoints tested successfully. Screenshot 2024-08-06 at 8 36 10 pm

@sammyrolfe  - this PR has been updated and merge conflict resolved - can you please review this one more time?

File: db/schema.rb (hunk 1 above)
Comment (DudeUnleashed) [reply db/schema.rb:47]: Originally there were uft8 as the charset, sql then changed the old entries to uft8mb3, this was as it is outdated and they wish to use the uft8 shorthand elsewhere for their new versions. Any new table generated now uses the uft8mb4 charset

**Comment by aNebula** [comment]
@DudeUnleashed are these already included in the PRs that came next? Otherwise please open upstream PR against https://github.com/doubtfire-lms/doubtfire-api/tree/new/course-flow.

**Pull request closed without merging**